src/Native/Runtime/thread.cpp

```diff
@@ -1118,7 +1121,15 @@ FORCEINLINE bool Thread::InlineTryFastReversePInvoke(ReversePInvokeFrame * pFram
 void Thread::ReversePInvokeAttachOrTrapThread(ReversePInvokeFrame * pFrame)
 {
     if (!IsStateSet(TSF_Attached))
+    {
+        if (*InitializeRuntimePtr != NULL && RuntimeInitializingThread != this)
```

Comment:
This is not thread safe. If two threads happen to be first to call a managed method at the same time, both of them will start initializing the runtime and bad things will happen. There needs to be logic to wait for the initialization to finish that will kick in for the second thread. A spin wait loop is probably the easiest way to do that.

Reply:
There is problem even without the race actually. The `RuntimeInitializingThread != this` condition will be false for the second thread calling here and the initialization will be done second time.